File: frontend/src/services/recipes.js
import axios from "axios";

const baseUrl = `${BACKEND_URL}/recipes`;

const getAll = async () => {
  const response = await axios.get(baseUrl);
  return response.data;
};

const searchByIngredients = async ({ keywords, offset }) => {
  const stringKeywords = keywords.join("&keywords=");
  const queryString = `?offset=${offset}&keywords=${stringKeywords}`;
  const response = await axios.get(`${baseUrl}/ingredients${queryString}`);
  return response.data;
};

const searchByTitle = async ({ keywords, offset }) => {
  const stringKeywords = keywords.join("&keywords=");
  const queryString = `?offset=${offset}&keywords=${stringKeywords}`;
  const response = await axios.get(`${baseUrl}/title${queryString}`);
  return response.data;
};

const services = { getAll, searchByIngredients, searchByTitle };

export default services;

File: backend/application/routes/recipes.py
from flask import Blueprint, jsonify, request
from application.utils.error_handler import ErrorHandler
from application.repo.csv.recipe.searchsorter import SearchSorter
from application.repo.csv.csv_repo import file_to_list_of_dict
blueprint = Blueprint("recipes", __name__)

error_handler = ErrorHandler(blueprint)
recipe_list = file_to_list_of_dict()


@blueprint.route("/ingredients")
def search_by_ingredients():
    offset = int(request.args.get('offset', 0))
    batch = 20

    keywords = set(request.args.getlist('keywords'))
    search_sorter = SearchSorter(recipe_list=recipe_list)
    result = search_sorter.search_by_ingredients(
        keywords)[offset:offset + batch]
    return jsonify(result)


@blueprint.route("/title")
def search_by_title():
    offset = int(request.args.get('offset', 0))
    batch = 20

    keywords = set(request.args.getlist('keywords'))
    search_sorter = SearchSorter(recipe_list=recipe_list)
    result = search_sorter.search_by_title(
        keywords)[offset:offset + batch]
    return jsonify(result)
